src/Component/Generator/PostgresReference.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\Component\Generator;
+
+use Sonata\Component\Invoice\InvoiceInterface;
+use Sonata\Component\Order\OrderInterface;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+
+final class PostgresReference implements ReferenceInterface
+{
+    /**
+     * @var RegistryInterface
+     */
+    private $registry;
+
+    public function __construct(RegistryInterface $registry)
+    {
+        $this->registry = $registry;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function invoice(InvoiceInterface $invoice)
+    {
+        if (!$invoice->getId()) {
+            throw new \RuntimeException('The invoice is not persisted into the database');
+        }
+
+        $this->generateReference($invoice,
+            $this->registry->getManager()->getClassMetadata(get_class($invoice))->table['name']);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function order(OrderInterface $order)
+    {
+        if (!$order->getId()) {
+            throw new \RuntimeException('The order is not persisted into the database');
+        }
+
+        $this->generateReference($order,
+            $this->registry->getManager()->getClassMetadata(get_class($order))->table['name']);
+    }
+
+    /**
+     * @param mixed  $object
+     * @param string $tableName
+     *
+     * @throws \Exception
+     *
+     * @return string
+     */
+    protected function generateReference($object, $tableName)
+    {
+        $date = new \DateTime();
+
+        $sql = sprintf(
+            "SELECT count(id) as counter FROM %s WHERE created_at >= '%s' AND reference IS NOT NULL",
+            $tableName,
+            $date->format('Y-m-d')
+        );
+
+        $this->registry->getConnection()->exec(sprintf('BEGIN WORK'));
+
+        $this->registry->getConnection()->exec(sprintf('LOCK TABLE %s IN EXCLUSIVE MODE', $tableName));
+
+        try {
+            $statement = $this->registry->getConnection()->query($sql);
+            $row = $statement->fetch();
+
+            $reference = sprintf('%02d%02d%02d%06d',
+                $date->format('y'),
+                $date->format('n'),
+                $date->format('j'),
+                $row['counter'] + 1
+            );
+
+            $this->registry->getConnection()->update(
+                $tableName,
+                array('reference' => $reference),
+                array('id' => $object->getId())
+            );
+            $object->setReference($reference);
+        } catch (\Exception $e) {
+            $this->registry->getConnection()->exec(sprintf('ROLLBACK WORK'));
+
+            throw $e;
+        }
+
+        $this->registry->getConnection()->exec(sprintf('COMMIT WORK'));
```

Comment:
When accessing db through `getConnection()` - we need to commit manually.
So we start transaction with `BEGIN WORK`, roll back if something went wrong (`ROLLBACK WORK`) and commit changes (`COMMIT WORK`).
I'm pretty sure, that we can do all this with doctrine entities and ` $this->registry->getEntityManager()->lock()`, but then we should change `sonata.payment.generator.mysql` to something else. I suppose it's better to do it in the next major.